File: leetcode/algorithms/206_reverse_linked_list/main.py
from typing import Optional, List

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class ReverseLinkedList:
    ##
    # Iteration
    # - Time Complexity: O(N)
    # - Space Complexity: O(N)
    #
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head == None or head.next == None:
            return head

        stack: List = []

        while head != None:
            stack.append(head)
            head = head.next

        new_head: Optional[ListNode] = stack.pop()
        curr_head: Optional[ListNode] = new_head

        while len(stack) > 0:
            curr_head.next = stack.pop()
            curr_head = curr_head.next

        curr_head.next = None

        return new_head


    # Solution
    ##
    # Solution 1
    #
    # Iteration
    # - Time Complexity: O(N)
    # - Space Complexity: O(1)
    #
    def solution1(self, head: Optional[ListNode]) -> Optional[ListNode]:
        prev = None
        current = head

        while current:
            next_node = current.next
            current.next = prev
            prev = current
            current = next_node

        return prev

    ##
    # Solution 2
    #
    # Recursion
    # - Time Complexity: O(N)
    # - Space Complexity: O(N)
    #
    def solution2(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head or not head.next:
            return head

        new_head = self.solution2(head.next)

        head.next.next = head
        head.next = None

        return new_head

File: leetcode/algorithms/206_reverse_linked_list/main.java
import java.util.Stack;

class ReverseLinkedList {
  class ListNode {
    int val;
    ListNode next;

    ListNode() {
    }

    ListNode(int val) {
      this.val = val;
    }

    ListNode(int val, ListNode next) {
      this.val = val;
      this.next = next;
    }
  }

  /**
   * Iteration
   * - Time Complexity: O(N)
   * - Space Complexity: O(N)
   */
  public ListNode reverseList(ListNode head) {
    if (head == null || head.next == null) {
      return head;
    }

    Stack<ListNode> stack = new Stack<>();

    while (head != null) {
      stack.push(head);
      head = head.next;
    }

    ListNode newHead = stack.pop();
    ListNode currHead = newHead;

    while (!stack.empty()) {
      currHead.next = stack.pop();
      currHead = currHead.next;
    }

    currHead.next = null;

    return newHead;
  }


  // Solution
  /**
   * Solution 1
   * 
   * Iteration
   * - Time Complexity: O(N)
   * - Space Complexity: O(1)
   */
  public ListNode solution1(ListNode head) {
    ListNode prev = null;
    ListNode current = head;

    while (current != null) {
      ListNode nextTemp = current.next;
      current.next = prev;
      prev = current;
      current = nextTemp;
    }

    return prev;
  }

  /**
   * Solution 2
   * 
   * Recursion
   * - Time Complexity: O(N)
   * - Space Complexity: O(N)
   */
  public ListNode solution2(ListNode head) {
    if (head == null || head.next == null) {
      return head;
    }

    ListNode newHead = solution2(head.next);

    head.next.next = head;
    head.next = null;

    return newHead;
  }
}
